flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyTopNFunction.java

```diff
@@ -210,14 +210,19 @@ private void processElementWithoutRowNumber(RowData input, Collector<RowData> ou
 			RowData lastKey = lastEntry.getKey();
 			List<RowData> lastList = (List<RowData>) lastEntry.getValue();
 			// remove last one
-			RowData lastElement = lastList.remove(lastList.size() - 1);
-			if (lastList.isEmpty()) {
+			int size = lastList.size();
+			RowData lastElement = null;
+			if (size > 0) {
+				lastElement = lastList.get(size - 1);
```

Comment:
```suggestion
				lastElement = lastList.remove(size - 1);
```

Reply:
Remove on the list use index will be more efficient than `buffer.remove`.

Reply:
Good point. I just find that maybe we can use `TopNBuffer#removeLast` directly here.

Reply:
Good idea, I don't want to break the atomicity of `TopNBuffer`,  so I don't want to add new method which can operate currentTopNum directly in this class,  prefer to use `TopNBuffer#removeLast`

Reply:
But, I find `TopNBuffer#removeLast` also iterate the list, the efficient problem also exist.

Reply:
Yes. I think we should improve `TopNBuffer#removeLast`, we can call `remove(index)` and `get(index)` when it is a `List`.